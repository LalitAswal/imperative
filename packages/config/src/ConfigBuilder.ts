/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { IImperativeConfig } from "../../imperative";
import { Config } from "./Config";
import { IConfig } from "./doc/IConfig";
import { IConfigBuilderOpts } from "./doc/IConfigBuilderOpts";
import { IConfigProfile } from "./doc/IConfigProfile";

export class ConfigBuilder {
    private static readonly DEFAULT_ROOT_PROFILE_NAME = "my_profiles";

    /**
     * Build a new Config object from an Imperative CLI app configuration.
     * @param impConfig The Imperative CLI app configuration.
     * @param opts Options to control aspects of the builder.
     */
    public static async build(impConfig: IImperativeConfig, opts?: IConfigBuilderOpts): Promise<IConfig> {
        opts = opts || {};
        const config: IConfig = Config.empty();

        const baseProfileType: string = impConfig.baseProfile?.type;
        const rootProfileName: string = impConfig.templateProfileName || ConfigBuilder.DEFAULT_ROOT_PROFILE_NAME;

        for (const profile of impConfig.profiles) {
            let profileShortPath = `my_${profile.type}`;
            let profileLongPath = `profiles.${profileShortPath}`;
            if (baseProfileType && profile.type !== baseProfileType) {
                // Path should have two levels for non-base profiles
                profileShortPath = `${rootProfileName}.${profile.type}`;
                profileLongPath = `profiles.${rootProfileName}.profiles.${profile.type}`;
                lodash.set(config.profiles, `${rootProfileName}.properties`, {});
            }

            const properties: { [key: string]: any } = {};
            for (const [k, v] of Object.entries(profile.schema.properties)) {
                if (opts.populateProperties && v.includeInTemplate) {
                    if (v.secure) {
                        const propPath = `${profileLongPath}.properties.${k}`;
                        config.secure.push(propPath);

                        const propValue = opts.getSecureValue ? await opts.getSecureValue(k, v) : null;
                        if (propValue != null) {
                            // Save this value to be stored securely after profile is built
                            properties[k] = propValue;
                        }
                    } else {
                        if ((v as any).optionDefinition != null) {
                            // Use default value of ICommandOptionDefinition if present
                            properties[k] = (v as any).optionDefinition.defaultValue;
                        }
                        if (properties[k] === undefined) {
                            // Fall back to an empty value
                            properties[k] = this.getDefaultValue(v.type);
                        }
                    }
                }
            }

            // Add the profile to config and set it as default
            lodash.set(config, profileLongPath, {
                type: profile.type,
                properties
            });

            if (opts.populateProperties) {
                config.defaults[profile.type] = profileShortPath;
            }
        }

        // Hoist duplicate default properties
        if (config.profiles != null && config.profiles[rootProfileName] != null) {
            config.profiles[rootProfileName] = this.hoistTemplateProperties(config.profiles[rootProfileName]);
        }

        return config;
    }

    /**
     * Returns empty value that is appropriate for the property type.
     * @param propType The type of profile property
     * @returns Null or empty object
     */
    private static getDefaultValue(propType: string | string[]): any {
        // TODO How to handle profile property with multiple types
        if (Array.isArray(propType)) {
            propType = propType[0];
        }
        switch (propType) {
            case "string":  return "";
            case "number":  return 0;
            case "object":  return {};
            case "array":   return [];
            case "boolean": return false;
            default:        return null;
        }
    }

    /**
     * Moves duplicate default properties across child profiles up to root
     * profile.
     * @param rootProfile The root profile object of the config JSON
     * @returns The root profile with duplicate properties hoisted
     */
    private static hoistTemplateProperties(rootProfile: IConfigProfile): IConfigProfile {
        // Flatten properties into object that maps property name to list of values
        const flattenedProps: { [key: string]: any[] } = {};
        for (const childProfile of Object.values(rootProfile.profiles)) {
            for (const [k, v] of Object.entries(childProfile.properties)) {
                flattenedProps[k] = [...(flattenedProps[k] || []), v];
            }
        }
        // List property names defined multiple times with the same value
        const duplicateProps: string[] = [];
        for (const [k, v] of Object.entries(flattenedProps)) {
            if (v.length > 1 && (new Set(v)).size === 1) {
                duplicateProps.push(k);
            }
        }
        // Remove duplicate properties from child profiles and store them in root profile
        for (const propName of duplicateProps) {
            rootProfile.properties[propName] = flattenedProps[propName][0];
            for (const childProfile of Object.values(rootProfile.profiles)) {
                delete childProfile.properties[propName];
            }
        }
        return rootProfile;
    }
}
