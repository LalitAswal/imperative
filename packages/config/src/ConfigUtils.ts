/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeConfig } from "../../utilities";
import { ImperativeError } from "../../error";

export function secureSaveError(solution?: string): ImperativeError {
    const displayName = ImperativeConfig.instance.loadedConfig.productDisplayName || ImperativeConfig.instance.loadedConfig.name;
    let details = `Possible Solutions:\n` +
        ` 1. Reinstall ${displayName}. On Linux systems, also make sure to install the prerequisites listed in ${displayName} documentation.\n` +
        ` 2. Ensure ${displayName} can access secure credential storage. ${displayName} needs access to the OS to securely save credentials.`;
    if (solution != null) {
        details += `\n 3. ${solution}`;
    }
    return new ImperativeError({
        msg: "Unable to securely save credentials.",
        additionalDetails: details
    });
}
