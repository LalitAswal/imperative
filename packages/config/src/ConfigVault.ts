/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IConfigVault } from "./doc/IConfigVault";

export class ConfigVault {
    private _vault: IConfigVault;

    // Maximum credential length allowed on Windows 7 and newer
    private static readonly MAX_CREDENTIAL_LENGTH = 2560;

    /**
     * Initialize a config vault using the provided vault backend.
     * @param vault Backend with load and save methods
     */
    constructor(vault: IConfigVault) {
        this._vault = vault;
    }

    public async load(key: string): Promise<any> {
        // Load initial value from vault
        let value = await this._vault.load(key);

        // On Windows, try to load more values from vault
        if (process.platform === "win32" && value != null) {
            let index = 1;
            // Check if we've finished reading null-terminated JSON string
            while (!value.endsWith('\0')) {
                index++;
                // Load more values from vault and concat them
                value += await this._vault.load(`${key}-${index}`);
            }
            // Strip off trailing null char
            value = value.slice(0, -1);
        }

        return value;
    }

    public async save(key: string, value: string): Promise<void> {
        // Calculate max credential length before Base64 encoding
        const maxCredentialLength = Math.ceil(ConfigVault.MAX_CREDENTIAL_LENGTH * 0.75) - 4;
        if (process.platform === "win32" && value.length > maxCredentialLength) {
            value += '\0';
            let index = 1;
            while (value.length > 0) {
                const newKey = (index > 1) ? `${key}-${index}` : key;
                await this._vault.save(newKey, value.slice(0, maxCredentialLength));
                value = value.slice(maxCredentialLength);
                index++;
            }
        } else {
            await this._vault.save(key, value);
        }
    }
}
