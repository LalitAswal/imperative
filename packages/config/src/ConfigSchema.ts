/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";
import { IExplanationMap, TextUtils } from "../../utilities/src/TextUtils";
import { ICommandProfileProperty } from "../../cmd";
import { IProfileProperty, IProfileSchema, IProfileTypeConfiguration } from "../../profiles";
import { IConfigSchema, IConfigUpdateSchemaOptions } from "./doc/IConfigSchema";
import { ImperativeConfig } from "../../utilities/src/ImperativeConfig";
import { Logger } from "../../logger/src/Logger";
import { Config } from "./Config";
import { ImperativeError } from "../../error/src/ImperativeError";
export class ConfigSchema {
    /**
     * JSON schema URI stored in $schema property of the schema
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly JSON_SCHEMA = "https://json-schema.org/draft/2020-12/schema";

    /**
     * Version number stored in $version property of the schema
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly SCHEMA_VERSION = 3;

    /**
     * Pretty explanation of the schema objects
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly explainSchemaSummary: IExplanationMap = {
        $schema: "URL",
        $version: "Version",
        properties: {
            defaults: "Default Definitions",
            explainedParentKey: "Properties",
            ignoredKeys: null
        },
        explainedParentKey: "Schema",
        ignoredKeys: null
    };

    /**
     * Transform an Imperative profile schema to a JSON schema. Removes any
     * non-JSON-schema properties and translates anything useful.
     * @param schema The Imperative profile schema
     * @returns JSON schema for profile properties
     */
    private static generateSchema(schema: IProfileSchema): any {
        const properties: { [key: string]: any } = {};
        const secureProps: string[] = [];
        for (const [k, v] of Object.entries(schema.properties || {})) {
            properties[k] = { type: v.type };
            const cmdProp = v as ICommandProfileProperty;
            if (cmdProp.optionDefinition != null) {
                properties[k].description = cmdProp.optionDefinition.description;
                if (cmdProp.optionDefinition.defaultValue != null) {
                    properties[k].default = cmdProp.optionDefinition.defaultValue;
                }
                if (cmdProp.optionDefinition.allowableValues != null) {
                    properties[k].enum = cmdProp.optionDefinition.allowableValues.values;
                }
            }
            if (v.secure) {
                secureProps.push(k);
            }
        }

        const propertiesSchema: any = {
            type: schema.type,
            title: schema.title,
            description: schema.description,
            properties
        };
        if (schema.required) {
            propertiesSchema.required = schema.required;
        }

        const secureSchema: any = {
            prefixItems: {
                enum: secureProps
            }
        };

        if (secureProps.length > 0) {
            return { properties: propertiesSchema, secure: secureSchema };
        } else {
            return { properties: propertiesSchema };
        }
    }

    /**
     * Transform a JSON schema to an Imperative profile schema.
     * @param schema The JSON schema for profile properties
     * @returns Imperative profile schema
     */
    private static parseSchema(schema: any): IProfileSchema {
        const properties: { [key: string]: IProfileProperty } = {};
        for (const [k, v] of Object.entries((schema.properties.properties || {}) as { [key: string]: any })) {
            properties[k] = { type: v.type };
            if (schema.secure?.prefixItems.enum.includes(k)) {
                properties[k].secure = true;
            }
            if (v.description != null || v.default != null || v.enum != null) {
                (properties[k] as ICommandProfileProperty).optionDefinition = {
                    name: k,
                    type: v.type,
                    description: v.description,
                    defaultValue: v.default,
                    allowableValues: v.enum ? { values: v.enum } : undefined
                };
            }
        }

        return {
            title: schema.properties.title,
            description: schema.properties.description,
            type: schema.properties.type,
            properties,
            required: schema.properties.required
        };
    }

    /**
     * Dynamically builds the config schema for this CLI.
     * @param profiles The profiles supported by this CLI
     * @returns JSON schema for all supported profile types
     */
    public static buildSchema(profiles: IProfileTypeConfiguration[]): IConfigSchema {
        const entries: any[] = [];
        const defaultProperties: { [key: string]: any } = {};
        profiles.forEach((profile: { type: string, schema: IProfileSchema }) => {
            entries.push({
                if: { properties: { type: { const: profile.type } } },
                then: { properties: this.generateSchema(profile.schema) }
            });
            defaultProperties[profile.type] = { type: "string" };
        });
        return {
            $schema: ConfigSchema.JSON_SCHEMA,
            $version: ConfigSchema.SCHEMA_VERSION,
            type: "object",
            description: "config",
            properties: {
                profiles: {
                    type: "object",
                    description: "named profiles config",
                    patternProperties: {
                        "^\\S*$": {
                            type: "object",
                            description: "a profile",
                            properties: {
                                type: {
                                    description: "the profile type",
                                    type: "string"
                                },
                                properties: {
                                    description: "the profile properties",
                                    type: "object"
                                },
                                profiles: {
                                    description: "additional sub-profiles",
                                    type: "object",
                                    $ref: "#/properties/profiles"
                                },
                                secure: {
                                    description: "secure property names",
                                    type: "array",
                                    prefixItems: {
                                        type: "string"
                                    },
                                    uniqueItems: true
                                }
                            },
                            allOf: entries
                        }
                    }
                },
                defaults: {
                    type: "object",
                    description: "default profiles config",
                    properties: defaultProperties
                }
            }
        };
    }

    /**
     * Loads Imperative profile schema objects from a schema JSON file.
     * @param schema The schema JSON for config
     */
    public static loadSchema(schema: IConfigSchema): IProfileTypeConfiguration[] {
        const patternName = Object.keys(schema.properties.profiles.patternProperties)[0];
        const profileSchemas: IProfileTypeConfiguration[] = [];
        for (const obj of schema.properties.profiles.patternProperties[patternName].allOf) {
            profileSchemas.push({
                type: obj.if.properties.type.const,
                schema: this.parseSchema(obj.then.properties)
            });
        }
        return profileSchemas;
    }

    /**
     * Updates Imperative Config Schema objects from a schema JSON file.
     * @param options        The options object
     * @param options.layer  The layer in which we should update the schema file(s). Defaults to the active layer.
     * @param options.schema The optional schema object to use. If not provided, we build the schema object based on loadedConfig.profiles
     * @returns List of updated paths with the new/loaded or given schema
     */
    public static updateSchema(options?: IConfigUpdateSchemaOptions): { [key: string]: { schema: string, updated: boolean } } {
        const opts: IConfigUpdateSchemaOptions = { ...{ layer: "active", depth: 0 }, ...(options ?? {}) };
        const schema = opts.schema ?? ConfigSchema.buildSchema(ImperativeConfig.instance.loadedConfig.profiles);
        const config = ImperativeConfig.instance.config;
        const initialLayer = config.api.layers.get();
        let updatedPaths: { [key: string]: { schema: string, updated: boolean } } = {};
        switch (opts.layer) {
            case "active": {
                Logger.getAppLogger().debug(`Updating "${initialLayer.path}" with: \n` +
                    TextUtils.prettyJson(TextUtils.explainObject(schema, ConfigSchema.explainSchemaSummary, false), null, false));
                config.setSchema(schema);
                const schemaInfo = config.getSchemaInfo();
                updatedPaths = { [initialLayer.path]: { schema: schemaInfo?.original, updated: schemaInfo?.local } };
                break;
            }
            case "global": {
                config.api.layers.activate(false, true);
                updatedPaths = { ...updatedPaths, ...ConfigSchema.updateSchema({ schema }) };

                if (config.api.layers.exists(true, true)) {
                    config.api.layers.activate(true, true);
                    updatedPaths = { ...updatedPaths, ...ConfigSchema.updateSchema({ schema }) };
                }

                config.api.layers.activate(initialLayer.user, initialLayer.global, initialLayer.path);
                break;
            }
            case "all": {
                let currentLayer = initialLayer;
                let nextSchemaLocation = initialLayer.path;
                while (nextSchemaLocation != null) {
                    updatedPaths = { ...updatedPaths, ...ConfigSchema.updateSchema({ schema }) };

                    // Check if we are in a user config
                    if (!currentLayer.user) {
                        // If we are not in a user config, we can move on to the next directory
                        nextSchemaLocation = Config.search(config.schemaName, { startDir: path.join(path.dirname(currentLayer.path), "..") });
                    }
                    if (nextSchemaLocation != null) {
                        config.api.layers.activate(false, false, nextSchemaLocation);
                    }
                    currentLayer = config.api.layers.get();
                }

                if (!initialLayer.global) {
                    // Do not update the global layer if that's where we started from
                    updatedPaths = { ...updatedPaths, ...ConfigSchema.updateSchema({ layer: "global", schema }) };
                }

                /**
                 * Method: `**`
                 * Result: DO NOT USE THIS
                 BIG NO-NO: Takes about 15 seconds from /root
                 Results from / ; time: 42.898s
                    const matches = glob.sync(`**\/${config.schemaName}`, {}).filter((match: string) => {
                        return match.split("/").length <= opts.depth + 1;
                    });
                */

                /**
                 * Method: `* /* /* /* ...`
                 * Result: DO NOT USE THIS
                 * Takes about a second from /root
                 * Similar to `**`: Takes about 50 seconds from /
                    let matches = glob.sync(`./${config.schemaName}`, {});
                    let str = "*\/"
                    for (let index = 0; index < opts.depth - 1; index++) {
                        str += "*\/"
                        matches = matches.concat(glob.sync(`${str}${config.schemaName}`, {}))
                    }
                 */

                /**
                 * Method: `fast-glob`
                 * Result: Best so far : )
                 * Takes about a second from /root
                 * And less than 20 seconds from /
                    const fg = require("fast-glob");
                    const matches = fg.sync(`**\/${config.schemaName}`, { onlyFiles: true, deep: opts.depth + 1 });
                 */

                if (opts.depth > 0) {
                    const fg = require("fast-glob");
                    // The `cwd` does not participate in Fast-glob's depth calculation, hence the + 1
                    const matches: string[] = fg.sync(`**/${config.schemaName}`, { dot: true, onlyFiles: true, deep: opts.depth + 1});
                    const globalProjConfig = config.findLayer(false, true);
                    const globalUserConfig = config.findLayer(true, true);
                    matches.forEach(schemaLoc => {
                        if (config.api.layers.exists(false, false, schemaLoc)) {
                            config.api.layers.activate(false, false, schemaLoc);
                            const layer = config.layerActive();
                            // NOTE: Configs are assumed to be always local (because of path.resolve(layer.path)),
                            //       if we want to support Config URLs here, we need to call the config import APIs
                            if (path.resolve(layer.path) !== globalProjConfig.path && path.resolve(layer.path) !== globalUserConfig.path) {
                                updatedPaths = { ...updatedPaths, ...ConfigSchema.updateSchema({ schema }) };
                            }
                        }
                    });
                }

                // Back to initial layer
                config.api.layers.activate(initialLayer.user, initialLayer.global, initialLayer.path);
                break;
            }
            default: {
                throw new ImperativeError({
                    msg: "Unrecognized layer parameter for ConfigSchema.updateSchemas"
                });
            }
        }
        return updatedPaths;
    }
}
