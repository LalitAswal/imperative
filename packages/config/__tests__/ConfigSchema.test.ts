/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IProfileTypeConfiguration } from "../../profiles";
import { ConfigSchema } from "../src/ConfigSchema";
import { cloneDeep } from "lodash";


describe("Config Schema", () => {
    const schema = ConfigSchema;
    const testProfileConfiguration: IProfileTypeConfiguration[] = [
    {
        type: "zosmf",
        schema: {
            title: "zosmf",
            description: "A fake zosmf profile",
            type: "zosmf",
            required: [],
            properties: {
                host: {
                    type: "string",
                    secure: false
                }
            }
        }
    },
    {
        type: "base",
        schema: {
            title: "base",
            description: "A fake base profile",
            type: "base",
            required: [],
            properties: {
                port: {
                    type: "number",
                    secure: false
                }
            }
        }
    }];
    const testProfileConfigurationSecure: IProfileTypeConfiguration[] = [
    {
        type: "zosmf",
        schema: {
            title: "zosmf",
            description: "A fake zosmf profile",
            type: "zosmf",
            required: [],
            properties: {
                host: {
                    type: "string",
                    secure: true
                }
            }
        }
    }];
    // TODO Why doesn't type work here when optionDefinition is defined
    const testProfileConfigurationOptionDefinition: any = [
        {
            type: "zosmf",
            schema: {
                title: "zosmf",
                description: "A fake zosmf profile",
                type: "zosmf",
                required: [],
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["h"],
                            description: "The fake host to connect to",
                            type: "string",
                            defaultValue: "fake"
                        }
                    }
                }
            }
        }];

    it("should be able to successfully build with no profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = [];
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any = [];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with a single profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfiguration);
        testConfig.pop();
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [{
            if: {
                properties: {
                    type: {
                        const: "zosmf"
                    }
                }
            },
            then: {
                properties: {
                    properties: {
                        description: "A fake zosmf profile",
                        properties: {
                            host: {
                                type: "string"
                            },
                        },
                        required: [],
                        title: "zosmf",
                        type: "zosmf"
                    }
                }
            }
        }];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with two profile type configurations", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfiguration);
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [
            {
                if: {
                    properties: {
                        type: {
                            const: "zosmf"
                        }
                    }
                },
                then: {
                    properties: {
                        properties: {
                            description: "A fake zosmf profile",
                            properties: {
                                host: {
                                    type: "string"
                                },
                            },
                            required: [],
                            title: "zosmf",
                            type: "zosmf"
                        }
                    }
                }
            },
            {
                if: {
                    properties: {
                        type: {
                            const: "base"
                        }
                    }
                },
                then: {
                    properties: {
                        properties: {
                            description: "A fake base profile",
                            properties: {
                                port: {
                                    type: "number"
                                },
                            },
                            required: [],
                            title: "base",
                            type: "base"
                        }
                    }
                }
            }
        ];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with a secure single profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfigurationSecure);
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [{
            if: {
                properties: {
                    type: {
                        const: "zosmf"
                    }
                }
            },
            then: {
                properties: {
                    properties: {
                        description: "A fake zosmf profile",
                        properties: {
                            host: {
                                type: "string",
                                secure: true
                            },
                        },
                        required: [],
                        title: "zosmf",
                        type: "zosmf"
                    }
                }
            }
        }];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with a secure single profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfigurationOptionDefinition);
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [{
            if: {
                properties: {
                    type: {
                        const: "zosmf"
                    }
                }
            },
            then: {
                properties: {
                    properties: {
                        description: "A fake zosmf profile",
                        properties: {
                            host: {
                                type: "string",
                                default: "fake",
                                description: "The fake host to connect to"
                            },
                        },
                        required: [],
                        title: "zosmf",
                        type: "zosmf"
                    }
                }
            }
        }];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });
});