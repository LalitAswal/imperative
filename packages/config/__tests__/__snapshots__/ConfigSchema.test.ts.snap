// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Schema should be able to successfully build with a secure single profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2019-09/schema#",
  "$version": 1,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "description": "the type",
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "secure": true,
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "zosmf",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
    "secure": Object {
      "description": "secure properties",
      "items": Object {
        "description": "path to a property",
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with a secure single profile type configuration 2`] = `
Object {
  "$schema": "https://json-schema.org/draft/2019-09/schema#",
  "$version": 1,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "description": "the type",
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "default": "fake",
                        "description": "The fake host to connect to",
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "zosmf",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
    "secure": Object {
      "description": "secure properties",
      "items": Object {
        "description": "path to a property",
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with a single profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2019-09/schema#",
  "$version": 1,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "description": "the type",
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "zosmf",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
    "secure": Object {
      "description": "secure properties",
      "items": Object {
        "description": "path to a property",
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with no profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2019-09/schema#",
  "$version": 1,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "description": "the type",
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
    "secure": Object {
      "description": "secure properties",
      "items": Object {
        "description": "path to a property",
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with two profile type configurations 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2019-09/schema#",
  "$version": 1,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "description": "the type",
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "zosmf",
                  },
                },
              },
            },
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "base",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake base profile",
                    "properties": Object {
                      "port": Object {
                        "type": "number",
                      },
                    },
                    "required": Array [],
                    "title": "base",
                    "type": "base",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
    "secure": Object {
      "description": "secure properties",
      "items": Object {
        "description": "path to a property",
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;
